modules/prediction/evaluator/cyclist/cyclist_keep_lane_evaluator.cc

```diff
@@ -0,0 +1,73 @@
+/******************************************************************************
+ * Copyright 2018 The Apollo Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *****************************************************************************/
+
+#include "modules/prediction/evaluator/cyclist/cyclist_keep_lane_evaluator.h"
+
+namespace apollo {
+namespace prediction {
+
+void CyclistKeepLaneEvaluator::Evaluate(Obstacle* obstacle_ptr) {
+  CHECK_NOTNULL(obstacle_ptr);
+  int id = obstacle_ptr->id();
+  if (!obstacle_ptr->latest_feature().IsInitialized()) {
+    AERROR << "Obstacle [" << id << "] has no latest feature.";
+    return;
+  }
+
+  Feature* latest_feature_ptr = obstacle_ptr->mutable_latest_feature();
+  CHECK_NOTNULL(latest_feature_ptr);
+  if (!latest_feature_ptr->has_lane() ||
+      !latest_feature_ptr->lane().has_lane_graph() ||
+      !latest_feature_ptr->lane().has_lane_feature()) {
+    ADEBUG << "Obstacle [" << id << "] has no lane graph.";
+    return;
+  }
+
+  LaneGraph* lane_graph_ptr =
+      latest_feature_ptr->mutable_lane()->mutable_lane_graph();
+  CHECK_NOTNULL(lane_graph_ptr);
+  if (lane_graph_ptr->lane_sequence_size() == 0) {
+    AERROR << "Obstacle [" << id << "] has no lane sequences.";
+    return;
+  }
+
+  std::string curr_lane_id =
+      latest_feature_ptr->lane().lane_feature().lane_id();
```

Comment:
In the current implementation, in some very occasional cases, it might be possible that lane_feature()'s lane_id is not used to construct lane_graph. This might result in all lanes having 0.0 probability. I don't know if it's better that, when this happens, we assign 1.0 to some arbitrary lane_sequence.

Reply:
In that case, the obstacle will not be regarded as "OnLane", thus it won't have lane feature. In obstacle.cc Line 961 and Line 969, lane id is guaranteed.